CRM/Case/Form/Activity/LinkCases.php

```diff
@@ -99,7 +99,11 @@ public static function formRule($values, $files, $form) {
     $errors = [];
 
     $linkCaseId = $values['link_to_case_id'] ?? NULL;
-    assert('is_numeric($linkCaseId)');
+
+    if (!CRM_Utils_Rule::numeric($linkCaseId)) {
```

Comment:
I think even the assert was wrong before because e.g. `-3.14` is not a valid case id but is numeric. (You can do this in the dev console with some javascript and then it fatals later with `INSERT INTO civicrm_case_activity (case_id , activity_id ) VALUES ( -3.14 , 57173 ) [nativecode=1264 ** Out of range value for column 'case_id' at row 1]`

CRM_Utils_Rule::positiveInteger oddly allows `0` so would also be wrong.

I think to be strictly correct it would be `if (is_numeric($linkCaseId) && ((int) $linkCaseId) <= 0) { $errors... }` to also deal with things like `true` and objects, although I'm not sure how you'd pass those via the form. So `if ((int) $linkCaseId <= 0) { $errors... }` is probably good enough.

Reply:
Just realizing my suggestion is wrong. How about `if (!CRM_Utils_Rule::positiveInteger($linkCaseId) || $linkCaseId == 0) { $errors ... }` ?

Reply:
@braders can you make that change?

Reply:
The last suggestion from @demeritcowboy looks sound. I've re-pushed.

Reply:
I probably should have run my suggestion before suggesting it. Doesn't quite work. Give me a minute.

Reply:
Two problems:

* All of the $errors indexes in that function are wrong. They "work" to validate the form but refer to a non-existent field name, so there's no error displayed. They should all be `$errors['link_to_case_id']`.
* `0` is tricky here. The validation comparison works, but then when the form reloads the entityref widget throws an api exception (apparently it's fine with -3.14, but doesn't like 0).
 
This works for me.

```patch
--- a/CRM/Case/Form/Activity/LinkCases.php
+++ b/CRM/Case/Form/Activity/LinkCases.php
@@ -99,15 +99,26 @@ class CRM_Case_Form_Activity_LinkCases {
     $errors = [];

     $linkCaseId = $values['link_to_case_id'] ?? NULL;
-    assert('is_numeric($linkCaseId)');
+
+    if (!CRM_Utils_Rule::positiveInteger($linkCaseId)) {
+      $errors['link_to_case_id'] = ts('The linked case ID is invalid.');
+    }
+
+    if ($linkCaseId == 0) {
+      // We can't just return $errors because when the page reloads the
+      // entityref widget throws an error before the page can display the error.
+      // It seems ok with other invalid values, just not 0.
+      CRM_Core_Error::statusBounce(ts('The linked case ID is invalid.'));
+    }
+
     if ($linkCaseId == CRM_Utils_Array::first($form->_caseId)) {
-      $errors['link_to_case'] = ts('Please select some other case to link.');
+      $errors['link_to_case_id'] = ts('Please select some other case to link.');
     }

     // do check for existing related cases.
     $relatedCases = $form->get('relatedCases');
     if (is_array($relatedCases) && array_key_exists($linkCaseId, $relatedCases)) {
-      $errors['link_to_case'] = ts('Selected case is already linked.');
+      $errors['link_to_case_id'] = ts('Selected case is already linked.');
     }

     return empty($errors) ? TRUE : $errors;
```

Or this variation, since any kind of non-numeric value is somebody fooling around and who doesn't deserve to see the form again...

```patch
--- a/CRM/Case/Form/Activity/LinkCases.php
+++ b/CRM/Case/Form/Activity/LinkCases.php
@@ -99,15 +99,26 @@ class CRM_Case_Form_Activity_LinkCases {
     $errors = [];

     $linkCaseId = $values['link_to_case_id'] ?? NULL;
-    assert('is_numeric($linkCaseId)');
+
+    if (!CRM_Utils_Rule::positiveInteger($linkCaseId) || $linkCaseId == 0) {
+      // We can't just return $errors because when the page reloads the
+      // entityref widget throws an error before the page can display the error.
+      // It seems ok with other invalid values, just not 0, but both are equally invalid.
+      CRM_Core_Error::statusBounce(ts('The linked case ID is invalid.'));
+    }
+
     if ($linkCaseId == CRM_Utils_Array::first($form->_caseId)) {
-      $errors['link_to_case'] = ts('Please select some other case to link.');
+      $errors['link_to_case_id'] = ts('Please select some other case to link.');
     }

     // do check for existing related cases.
     $relatedCases = $form->get('relatedCases');
     if (is_array($relatedCases) && array_key_exists($linkCaseId, $relatedCases)) {
-      $errors['link_to_case'] = ts('Selected case is already linked.');
+      $errors['link_to_case_id'] = ts('Selected case is already linked.');
     }

     return empty($errors) ? TRUE : $errors;
```

Reply:
Variation 2 seems best to me. Tested and re-pushed. Thanks for spotting these additional issues.

Reply:
Thanks!